ui/src/funding/contract.ts

```diff
@@ -27,31 +32,119 @@ const addresses = {
   switch (environment) {
     case ethereum.Environment.Local:
-      return addresses.pool.local;
+      return addresses.dai.local;
     case ethereum.Environment.Ropsten:
-      return addresses.pool.ropsten;
+      return addresses.dai.ropsten;
   }
 }
 
-export function pool(signer: Signer, address: string): PoolContract {
-  return new PoolContract(signer, address);
+export function daiToken(signer: Signer, address: string): ERC20 {
+  return Erc20Factory.connect(address, signer);
+}
+
+// Get the address of the Claims Contract for the given environment
+export function claimsAddress(environment: ethereum.Environment): string {
+  switch (environment) {
+    case ethereum.Environment.Local:
+      return addresses.claims.local;
+    case ethereum.Environment.Ropsten:
+      return addresses.claims.ropsten;
+  }
+}
+
+export function claims(signer: Signer, address: string): ClaimsContract {
+  return new ClaimsContract(signer, address);
+}
+
+export class ClaimsContract {
+  contract: Claims;
+
+  // The application binary interface (abi) of this contract,
+  // needed to decode transactions that it produces. May need
+  // to be updated if the API of this contract changes.
+  readonly abi = [
+    {
+      anonymous: false,
+      inputs: [
+        {
+          indexed: true,
+          internalType: "address",
+          name: "addr",
+          type: "address",
+        },
+      ],
+      name: "Claimed",
+      type: "event",
+    },
+    {
+      inputs: [
+        {
+          internalType: "uint256",
+          name: "format",
+          type: "uint256",
+        },
+        {
+          internalType: "bytes",
+          name: "payload",
+          type: "bytes",
+        },
+      ],
+      name: "claim",
+      outputs: [],
+      stateMutability: "nonpayable",
+      type: "function",
+    },
+  ];
+
+  constructor(signer: Signer, address: string) {
+    this.contract = ClaimsFactory.connect(address, signer);
+  }
+
+  async claim(root: string): Promise<void> {
+    const payload = ethers.utils.toUtf8Bytes(root);
+    return this.contract
+      .claim(0, payload)
+      .then((ctx: ContractTransaction) =>
+        transaction.add(transaction.claimRadicleIdentity(ctx, root))
+      );
+  }
+
+  // Fetch the latest Radicle Identity root claim by the given ethereum address.
+  // Return undefined if no claim is found.
+  async claimed(address: string): Promise<string | undefined> {
+    const filter = this.contract.filters.Claimed(address);
+    const events = await this.contract.queryFilter(filter, 0, "latest");
+    const last = events[events.length - 1];
+
+    if (last === undefined) {
+      return undefined;
+    }
+
+    const tx = await last.getTransaction();
+    const decoder = new InputDataDecoder(this.abi);
+    const decoded = decoder.decodeData(tx.data);
+    const payload = decoded.inputs[1];
+    const root = ethereum.bytesToString(payload);
```

Comment:
You could try the following
```suggestion
    const claimsFactory = new ClaimsFactory()
    const inputs = claimsFactory.interface.decodeFunctionData("claim", tx.data)
    const root = ethereum.bytesToString(inputs.payload);

```
This code may need some tweaking but it would eliminate the need to add the abi manually and eliminate the InputDataDecoder dependency.

Reply:
I did try that actually but `decodeFunctionData` does not return the inputs decoded and I am not sure how to proceed from therein. Not needing the abi manually, or being able to retrieve it programatically would absolutely be best.

Reply:
:eyes: I think I found a way

Reply:
Fixed in https://github.com/radicle-dev/radicle-upstream/pull/1726/commits/009d3a42654c1527f2828ee544c242807781f8e1